### PR TITLE
gb.cpp: optimize the channel update methods.

### PR DESCRIPTION
**Pull request by 057a3dd61f99517a3afea0051a49cb27994f94d**
I recently noticed that it takes very long to load a gba state, sometimes
dozens of seconds, and the older the state is, the longer it takes to load.

I found out it's caused by gameboy_sound_device does not save the m_last_updated
variable, and after loading a state, in the update_state method, now could end
up being much larger than m_last_updated, and leads to substantial yet wasted
calculation of the channel state.

Simply adding save_item(NAME(m_last_updated)) should fix the problem, however,
since it will break current existing saved states, and optimize the channel
update methods is in the TODO list, I took the latter choice, now the calculation
to update the channel state is really negligible and the state loading is
almost instantaneous.

### CODE REVIEW

File: src/devices/sound/gb.cpp [hunk 1]
```diff
@@ -351,27 +351,26 @@ void gameboy_sound_device::update_square_channel(struct SOUND &snd, uint64_t cyc
 			}
 		}
 
-		while (cycles > 0)
+		if (cycles & 3)
```
Comment (cuavas) [inline src/devices/sound/gb.cpp]: Why did you change the condition from `cycles < 4` to `cycles & 3` for this?  Is it absolutely guaranteed that cycles can never be a higher value, e.g. 6 or 7?

**Comment by 057a3dd61f99517a3afea0051a49cb27994f94d** [comment]
It could be a higher value, but that's not a problem. The original method repeatedly
subtracts cycles by 4 in the while loop, so when the condition cycles < 4 meets it
really means 1 <= cycles % 4 (the same as cycles & 3) <= 3, i.e. cycles & 3 is not
zero. This is calculated first since the while loop is eliminated to improve performance.

**Comment by DavidHaywood** [comment]
save state compatibility isn't really a concern either, correct code / implementation is more important.

probably not the most user-friendly of approaches, but maintaining compatibility with legacy states has never been a strong point of MAME anyway.

**Comment by rb6502** [comment]
Yeah, I'd like to see both actually; fix the save states by saving m_last_updated, but also optimize the update handler to reduce the CPU hit a little.
